source/RoslynAnalyzers/Testing/Integration/IntegrationTestForwardCancellationTokenToInvocationsAnalyzer.cs

```diff
@@ -0,0 +1,271 @@
+    /// For Octopus Integration Test, CancellationToken is defined in the base class.
+    /// The existing dotnet analyzer, CA2016, handle cases where CancellationToken is found on the method context / containing class, but
+    /// not when definition is in parent class. (IE: IntegrationTest)
+    /// This implementation is based (heavily) on CA2016 implementation. (https://github.com/dotnet/roslyn-analyzers/pull/3641)
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class IntegrationTestForwardCancellationTokenToInvocationsAnalyzer : DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            Descriptors.Oct2008IntegrationTestForwardCancellationTokenToInvocations()
+        );
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        void AnalyzeCompilationStart(CompilationStartAnalysisContext compilationStartAnalysisContext)
+        {
+            compilationStartAnalysisContext.RegisterOperationAction(
+                context =>
+                {
+                    if (!IsOperationOfInterest(context))
+                    {
+                        return;
+                    }
+
+                    var (cancellationTokenType, genericTaskType, genericValueTaskType) = GetWellKnownTypes(context);
+                    if (cancellationTokenType == null)
+                    {
+                        // Only ct type matters, for the generic type it is ok to be null (edge case).
+                        return;
+                    }
+
+                    var invocation = (IInvocationOperation)context.Operation;
+                    var method = invocation.TargetMethod;
+
+                    // Verify that the current invocation is not passing an explicit token already
+                    if (IsCancellationTokenAlreadyForwarded(invocation.Arguments, cancellationTokenType))
+                    {
+                        return;
+                    }
+
+                    // Verify method has optional ct, warn user to be explicit rather than relying on default.
+                    if (InvocationMethodTakesAToken(method, invocation.Arguments, cancellationTokenType) ||
+
+                        // Verify method has ct overload as last parameter.
+                        InvocationMethodHasCancellationTokenOverload(method, cancellationTokenType, genericTaskType, genericValueTaskType))
+                    {
+                        SyntaxNode? nodeToDiagnose = GetInvocationMethodNameNode(context.Operation.Syntax) ?? context.Operation.Syntax;
```

Comment:
nit - the compiler was probably hinting to use `var` as preferred here.

Reply:
I prefer to have explicit type when the variable type is not clear from the context. (IE: `rhs` operation or other prior routine / declaration)

Reply:
Fair enough. The only things to keep in mind though are:

- Consistency. Some variables are `var` while others are specific types. Which is which can be very subjective.
- PrettyBot would insist on `var` in other repos.

Reply:
Good point. They will be updated to `var` on the `fixer` work stream. Thanks for the pointer!

Reply:
Cheers - no need to go out of your way to change these ones, unless it's easy. It was a bit of a nit pick :)